File: 03_Rucksack_Reorganization/03.py
INPUT = "input"


def convert(rucksacks):
    converted_rucksacks = []
    for i in range(0, len(rucksacks), 3):
        converted_rucksacks.append(
            [
                rucksacks[i],
                rucksacks[i + 1],
                rucksacks[i + 2],
            ]
        )
    return converted_rucksacks


def get_priority(item):
    if item.islower():
        return ord(item) - ord("a") + 1
    else:
        return ord(item) - ord("A") + 27


def get_repeating_item(rucksack):
    N = len(rucksack)
    a = set(rucksack[: N // 2])
    b = set(rucksack[N // 2 :])
    return "".join(a.intersection(b))


def get_badge(rucksack):
    a = set(rucksack[0])
    b = set(rucksack[1])
    c = set(rucksack[2])

    a = a.intersection(b)
    a = a.intersection(c)

    return "".join(a)


def read_input():
    with open(INPUT, "r") as file:
        rucksacks = file.read().split()
    return rucksacks


def main():
    rucksacks = read_input()
    repeating_items = map(get_repeating_item, rucksacks)
    priorities_sum = sum(map(get_priority, repeating_items))
    print(priorities_sum)

    converted_rucksacks = convert(rucksacks)
    badges_sum = sum(map(get_priority, map(get_badge, converted_rucksacks)))
    print(badges_sum)


if __name__ == "__main__":
    assert get_repeating_item("vJrwpWtwJgWrhcsFMMfFFhFp") == "p"
    assert get_repeating_item("jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL") == "L"
    assert get_repeating_item("PmmdzqPrVvPwwTWBwg") == "P"

    assert get_priority("p") == 16
    assert get_priority("L") == 38
    assert get_priority("P") == 42
    assert get_priority("v") == 22
    assert get_priority("Z") == 52

    assert convert([1, 2, 3, 4, 5, 6, 7, 8, 9]) == [
        [1, 2, 3],
        [4, 5, 6],
        [7, 8, 9],
    ]

    assert (
        get_badge(
            [
                "vJrwpWtwJgWrhcsFMMfFFhFp",
                "jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL",
                "PmmdzqPrVvPwwTWBwg",
            ]
        )
        == "r"
    )

    main()

File: aoc_2022_elixir/03_Rucksack_Reorganization/03.exs
ExUnit.start()

defmodule AssertionTest do
  use ExUnit.Case, async: true

  test "the truth" do
    assert true
  end

  test "get_repeating_item" do
    assert Day03.get_repeating_item("qwertqs1df") == "q"
    assert Day03.get_repeating_item("vJrwpWtwJgWrhcsFMMfFFhFp") == "p"
    assert Day03.get_repeating_item("jqHRNqRjqzjGDLGLrsFMfFZSrLrFZsSL") == "L"
    assert Day03.get_repeating_item("PmmdzqPrVvPwwTWBwg") == "P"
  end

  test "get_priority" do
    assert Day03.get_priority("p") == 16
    assert Day03.get_priority("L") == 38
    assert Day03.get_priority("P") == 42
    assert Day03.get_priority("v") == 22
    assert Day03.get_priority("Z") == 52
  end
end

defmodule Day03 do
  def read_input() do
    {:ok, file} = File.read("./input")
    String.trim(file) |> String.split()
  end

  def get_repeating_item(rucksack) do
    [part_1, part_2] =
      String.codepoints(rucksack)
      |> Enum.chunk_every(div(String.length(rucksack), 2))

    MapSet.intersection(MapSet.new(part_1), MapSet.new(part_2))
    |> Enum.at(0)
  end

  def get_priority(item) do
    item_code = hd(String.to_charlist(item))

    cond do
      item_code >= ?a and item_code <= ?z -> item_code - ?a + 1
      item_code >= ?A and item_code <= ?Z -> item_code - ?A + 27
    end
  end
end

rucksacks = Day03.read_input()

rucksacks
|> Enum.map(&Day03.get_repeating_item/1)
|> Enum.map(&Day03.get_priority/1)
|> Enum.sum()
|> IO.inspect()
